### PR TITLE
completed l2pool2d

### PR DESCRIPTION
**Pull request by mei1127**
@BruceDai  @huningxin  @fdwr PTAL , thanks！

### CODE REVIEW

File: src/reduce.js [hunk 1]
```diff
@@ -143,6 +143,20 @@ export function reduceL1(input, options = {}) {
   return reduceSum(abs(input), options);
 }
 
+/* The l2 reducer */
+export function l2Reducer(previousValue, currentValue, currentIndex, array) {
+  if (currentIndex== 1) {
```
Comment (fdwr) [inline src/reduce.js:148]: [nit] `if (currentIndex == 1) {`

File: src/reduce.js [hunk 2]
```diff
@@ -143,6 +143,20 @@ export function reduceL1(input, options = {}) {
...
+    const sumOfSquares = previousValue*previousValue + currentValue * currentValue;
```
Comment (fdwr) [inline src/reduce.js:149]: [nit] `... = previousValue * previousValue ...`

File: test/l2pool2d_test.js [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+      23.302360395462088,
+      31.654383582688826,
+      34.51086785347479,
+    ]);
+  });
+
+  it('l2Pool2d nhwc', function() {
+    const x = new Tensor([1, 4, 4, 1], [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16]);
+    const windowDimensions = [3, 3];
+    const layout = 'nhwc';
```
Comment (fdwr) [inline test/l2pool2d_test.js:24]: (ignorable musing - no action expected) 🤔 I've often wondered why pooling doesn't have variations that pool across *batches* too, not only the "spatial" dimensions. e.g. One could generalize pooling consistently like reduction (which can apply from all dimensions to none), and pass windowDimensions of rank 4 with `[3, 1, 3, 3]` to pool along batches too, or just `[3, 1, 1, 1]` for only batches. Now, DML *can* achieve the latter via explicit strides (to map the first dimension to the last dimensions), but there's no way to access it via WebNN, and I've never heard of a need for it anyway. 🤷

File: src/reduce.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (mei1127) [reply src/reduce.js]: In this function, currentIndex starts from 1 by default, and the number with index=0 at the beginning defaults to previousValue. So I consider this situation: if (currentIndex== 1) {
     const sumOfSquares = previousValue * previousValue + currentValue * currentValue;}

E.g.
case:
const x = new Tensor([1, 1, 4, 4], [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16]) ;
     const windowDimensions = [3, 3];
log:
currentIndex: 1 previousValue: 1 currentValue: 2
currentIndex: 2 previousValue: 5 currentValue: 3
currentIndex: 3 previousValue: 14 currentValue: 5
currentIndex: 4 previousValue: 39 currentValue: 6
currentIndex: 5 previousValue: 75 currentValue: 7
currentIndex: 6 previousValue: 124 currentValue: 9
currentIndex: 7 previousValue: 205 currentValue: 10
currentIndex: 8 previousValue: 305 currentValue: 11

File: test/l2pool2d_test.js [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+    utils.checkValue(y, expected);
+  });
+
+  it('l2Pool2d autoPad same-upper nhwc', function() {
+    const x = new Tensor([1, 5, 5, 1], [
+      1,  2,  3,  4,  5,  6,  7,  8,  9,  10, 11, 12, 13,
+      14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25,
+    ]);
+    const windowDimensions = [5, 5];
+    const autoPad = 'same-upper';
```
Comment (fdwr) [inline test/l2pool2d_test.js:205]: Oh, we're still using `autoPad` for pooling.

@huningxin, you opened https://github.com/webmachinelearning/webnn/issues/326 where we talked about removing autoPad from conv2d, and @Honry removed it from the WebNN EP in https://github.com/microsoft/onnxruntime/pull/18688, but for consistency, wouldn't we want to be explicit with pooling too?

File: test/l2pool2d_test.js [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+    const y = l2Pool2d(x, {windowDimensions, strides, layout});
+    utils.checkShape(y, [1, 2, 2, 1]);
+    const expected = [
+      9.486832980505138,
+      13.038404810405298,
+      28.460498941515414,
+      32.4037034920393,
+    ];
+    utils.checkValue(y, expected);
+  });
```
Comment (fdwr) [inline test/l2pool2d_test.js:271]: Shall we add a single element test case here (e.g. `const x = new Tensor([1, 1, 1, 1]`), or wait for the WPT?

**Review by fdwr — APPROVED** [review]
I have a question, but assuming it's correct, and assuming we add a single element test case here or in the WPT, then LGTM.

**Review by fdwr — APPROVED** [review]
Thanks Mei.

File: src/reduce.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (mei1127) [reply src/reduce.js]: FYI: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/reduce
image

File: test/l2pool2d_test.js (hunk 5 above)
Comment (mei1127) [reply test/l2pool2d_test.js:284]: Okay,I have added a single element test case.

File: src/reduce.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (mei1127) [reply src/reduce.js]: image

**Review by fdwr — APPROVED** [review]

File: test/l2pool2d_test.js (hunk 4 above)
Comment (huningxin) [reply test/l2pool2d_test.js:226]: > but for consistency, wouldn't we want to be explicit with pooling too?

Yes, we should be consistent and only do explicit padding for pooling.

**Pull request merged**